### DISCUSSION TITLE
Proper use of InterfaceKernels

### DISCUSSION THREAD

**Discussion by ke7kto** [idaholab/moose#17628] [Q&A General] [answered]
I'm modeling a diffusion process with a vapor-solid interface and a temperature gradient in one dimension. This seems like a very good fit for `InterfaceKernels`, but I cannot obtain a steady-state solve using `InterfaceDiffusionBoundaryTerm`, `InterfaceDiffusionFluxMatch` or a combination of the two in any event such that concentrations are greater than zero at the interface.

Taking out the interface and setting making the variable representing molar quantity the same in each subdomain allows for a solve. The reaction_1D_steady test solves successfully on my installation.

I hate asking "what am I doing wrong?" questions, but I'm really stuck here. How do I use the InterfaceKernels properly? I have attached the input file I'm using.

[simple.txt](https://github.com/idaholab/moose/files/6327856/simple.txt)

**Comment by aeslaughter** [comment, marked as answer]
@lindsayad Can you help with this question?

**Reply by ke7kto** [reply]
It turns out that I have an issue which seems to be unrelated to InterfaceKernels. I'll update when I've isolated the root of the problem.

**Reply by ke7kto** [reply]
There are four issues with the input file that prevent convergence

1. The initial condition for temperature was not close enough to the solution
2. line 173 uses an expression that evaluates to log(0) at x=0
3. The boundary condition for line 122 is set to the wrong variable
4. There seems to be an issue with using 'exp' in parsedFunction to generate a material property. I can get convergence when I use a custom Material class using the same function

**Reply by lindsayad** [reply]
So are the `InterfaceKernels` you're using working as desired with the resolution of your issues 1-4?

**Reply by ke7kto** [reply]
Good question. They're not working as I would have expected them to, but I could be misinterpreting the way that they are supposed to work. Is the `InterfaceDiffusionBoundaryTerm` supposed to apply to diffusion within the interface, across the boundary, or both?

What I find most puzzling is that the observed concentration difference across the interface changes direction when I switch between radial and cartesian coordinates. I've made a new input file without the previous problems (and removing the temperature dependence), which is attached.
[simplenoT.txt](https://github.com/idaholab/moose/files/6353358/simplenoT.txt)

**Reply by ke7kto** [reply]
I'm not sure I was very clear. I do not think that the `InterfaceKernels` are working properly. Here's why:

The file I've attached has Dirichlet boundary conditions as follows:

1. Concentration 0.16 @ x=0
2. Concentration 0 @ x=0.031782

There is an interface at x=0.0061625. Everything is in 1D. I'm using `InterfaceDiffusionBoundaryTerm` and `InterfaceDiffusionFluxMatch`.

- Diffusivity is set to 10⁻⁵ throughout both subdomains. If `D` or `D_neighbor` are set to any value other than the diffusivity of the bulk, a no-flux solution is obtained. 

- For cartesian coordinates and boundary diffusivities of 10⁻⁵, there is a jump to a concentration of ~0.65 on the right-hand side of the interface, which seems to be in error.
- With `coord_type = RZ` and `rz_coord_axis = Y` (radially symmetric 1D diffusion) the results are more believable, with a concentration drop from ~0.055 to ~0.027 across the interface - though if I understand correctly, the concentrations on either side should be equal for this situation.

**Reply by ke7kto** [reply]
The solution is also dependent on the number of elements

**Reply by lindsayad** [reply]
Sorry for being slow to respond @ke7kto. I have to say that I've looked at this pair of interface kernels before and it took me a fair while to understand what they are doing in a finite element sense. And it's been long enough since then that I've forgotten again. @dschwen is the author of those objects so maybe he can explain them.

I will say that I am a fan of `PenaltyInterfaceDiffusion` which is actually a test object for the framework. I think we should probably move it into the framework so users have access to that object. But you are welcome to copy those sources for your own app. `PenaltyInterfaceDiffusion` simultaneously enforces equal flux across an interface and solution continuity as well. It also preserves the optimal convergence rate of your finite element method, e.g. second order if you are using first order Lagrange bases, given an appropriate choice of penalty parameter (for well scaled problems something like 1e6 is a good choice).

**Reply by ke7kto** [reply]
@lindsayad Thanks for the reply! Using `PenaltyInterfaceDiffusion`, I get a good solve, the results are in line with expectations. For my specific problem 1e6 was too large a penalty, values from 1.0-1e5 worked well. Is there a straightforward way I could check the matrix properties to get an idea of the appropriate scale for the penalty parameter in general?

Also, looking at `PenaltyInterfaceDiffusion`, It seems I can specify a material property to calculate what the discontinuity should be. Is there a way to define a material on an interface, especially as dependent on the difference of some variable on either side of the interface? I would need some functionality like that for the non-toy version of this problem.

**Reply by lindsayad** [reply]
> @lindsayad Thanks for the reply! Using `PenaltyInterfaceDiffusion`, I get a good solve, the results are in line with expectations. For my specific problem 1e6 was too large a penalty, values from 1.0-1e5 worked well. Is there a straightforward way I could check the matrix properties to get an idea of the appropriate scale for the penalty parameter in general?

It should at least be as large as your other diagonals, probably larger. It's a balance between how close to continuous you want the solution to be and how well conditioned you want your matrix to be. Matching the diagonals will give you the best conditioning but may not make your solution as close to continuous as you want (or maybe it does!). You can always inspect your matrix using the `-ksp_view_pmat` command line option (this is consumed by PETSc). You probably only want to use that option if your problem is reasonably small.

> Also, looking at `PenaltyInterfaceDiffusion`, It seems I can specify a material property to calculate what the discontinuity should be. Is there a way to define a material on an interface, especially as dependent on the difference of some variable on either side of the interface? I would need some functionality like that for the non-toy version of this problem.

You are looking for an `InterfaceMaterial`. You can find an example of just your use case in `moose/test/src/materials/JumpInterfaceMaterial.C`

**Reply by ke7kto** [reply]
This is exactly what I needed. Thank you.
